data-crawl/functions.py

```diff
@@ -0,0 +1,50 @@
+import variables
+import os
+import urllib.request
+import urllib.error
+from datetime import datetime
+import boto3
+
+def preparation(symbol_storagepath: str = variables.BASEPATH_SYMBOL, logfilepath: str = variables.DUCA_LOGFILE_PATH):
+    if not os.path.exists(logfilepath): 
+        with open(logfilepath, 'w'): pass
+    if not os.path.exists(symbol_storagepath):
+        os.mkdir(symbol_storagepath)
+    return {
+        'logfilepath': logfilepath,
+        'symbolpath': symbol_storagepath,
+        'utcnow': {
+            'year': datetime.utcnow().year,
+            'month': datetime.utcnow().month,
+            'day': datetime.utcnow().day,
+            'hour': datetime.utcnow().hour
+        }
+    }
+
+def load_ticks_hour_dukascopy(year:int, month:int, day:int, hour:int, symbol:str, symbol_storagepath: str = variables.BASEPATH_SYMBOL, logfilepath: str = variables.DUCA_LOGFILE_PATH):
+    url = f'{variables.DUCA_APIBASE}/{symbol.upper()}/{year:04d}/{(month-1):02d}/{day:02d}/{hour:02d}h_ticks.bi5'
+    filename = f'{year:04d}-{(month-1):02d}-{day:02d}-{hour:02d}h_ticks.bi5'
+    for i in range(5):
+        try:
+            urllib.request.urlretrieve(url, f'{symbol_storagepath}{filename}')
+        except urllib.error.HTTPError as e:
+            with open(logfilepath, 'a') as file:
+                file.write('ERROR: ' + str(e.status) + ' ' + url)
```

Comment:
do you really need to save it to a file?
File access is slow. If the date is not too huge then keep it in memory. Or write multiple files at once